File: src/device.rs
use std::fmt::Write as _;
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use rand::RngCore;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

const DEFAULT_CLIENT_INFO: &str =
    "<MacBookPro13,2> <macOS;13.1;22C65> <com.apple.AuthKit/1 (com.apple.dt.Xcode/3594.4.19)>";

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct DeviceData {
    #[serde(rename = "UUID")]
    pub unique_device_identifier: String,
    #[serde(rename = "clientInfo")]
    pub server_friendly_description: String,
    #[serde(rename = "identifier")]
    pub adi_identifier: String,
    #[serde(rename = "localUUID")]
    pub local_user_uuid: String,
}

#[derive(Debug, Clone)]
pub struct Device {
    path: PathBuf,
    pub data: DeviceData,
    pub initialized: bool,
}

impl Device {
    pub fn load(path: impl AsRef<Path>) -> Result<Self> {
        let path = path.as_ref().to_path_buf();

        if !path.exists() {
            return Ok(Self {
                path,
                data: DeviceData::default(),
                initialized: false,
            });
        }

        let bytes = fs::read(&path)
            .with_context(|| format!("failed to read device file {}", path.display()))?;
        let data: DeviceData = serde_json::from_slice(&bytes)
            .with_context(|| format!("failed to parse device file {}", path.display()))?;

        Ok(Self {
            path,
            data,
            initialized: true,
        })
    }

    pub fn initialize_defaults(&mut self) {
        self.data.server_friendly_description = DEFAULT_CLIENT_INFO.to_string();
        self.data.unique_device_identifier = Uuid::new_v4().to_string().to_uppercase();
        self.data.adi_identifier = random_hex(8, false);
        self.data.local_user_uuid = random_hex(32, true);
        self.initialized = true;
    }

    pub fn persist(&self) -> Result<()> {
        if let Some(parent) = self.path.parent() {
            fs::create_dir_all(parent)
                .with_context(|| format!("failed to create parent dir {}", parent.display()))?;
        }

        let bytes = serde_json::to_vec_pretty(&self.data)?;
        fs::write(&self.path, bytes)
            .with_context(|| format!("failed to write device file {}", self.path.display()))?;

        Ok(())
    }
}

fn random_hex(byte_len: usize, uppercase: bool) -> String {
    let mut bytes = vec![0_u8; byte_len];
    rand::thread_rng().fill_bytes(&mut bytes);

    let mut output = String::with_capacity(byte_len * 2);
    for byte in bytes {
        let _ = write!(output, "{byte:02x}");
    }

    if uppercase {
        output.make_ascii_uppercase();
    }

    output
}

File: js/src/device.ts
// Device identity management — loads or generates device.json

import type { AnisetteDeviceConfig, DeviceJson } from "./types.js";
import { randomHex, randomUUID } from "./utils.js";

const DEFAULT_CLIENT_INFO =
  "<MacBookPro13,2> <macOS;13.1;22C65> <com.apple.AuthKit/1 (com.apple.dt.Xcode/3594.4.19)>";

export class Device {
  readonly uniqueDeviceIdentifier: string;
  readonly serverFriendlyDescription: string;
  readonly adiIdentifier: string;
  readonly localUserUuid: string;

  private constructor(data: DeviceJson) {
    this.uniqueDeviceIdentifier = data.UUID;
    this.serverFriendlyDescription = data.clientInfo;
    this.adiIdentifier = data.identifier;
    this.localUserUuid = data.localUUID;
  }

  /** Load from a parsed device.json object, or generate defaults if null. */
  static fromJson(
    json: DeviceJson | null,
    overrides?: Partial<AnisetteDeviceConfig>
  ): Device {
    const defaults = Device.generateDefaults();
    const base: DeviceJson = json ?? {
      UUID: defaults.uniqueDeviceId,
      clientInfo: defaults.serverFriendlyDescription,
      identifier: defaults.adiId,
      localUUID: defaults.localUserUuid,
    };

    if (overrides) {
      if (overrides.uniqueDeviceId) base.UUID = overrides.uniqueDeviceId;
      if (overrides.serverFriendlyDescription)
        base.clientInfo = overrides.serverFriendlyDescription;
      if (overrides.adiId) base.identifier = overrides.adiId;
      if (overrides.localUserUuid) base.localUUID = overrides.localUserUuid;
    }

    return new Device(base);
  }

  /** Serialize back to the device.json wire format. */
  toJson(): DeviceJson {
    return {
      UUID: this.uniqueDeviceIdentifier,
      clientInfo: this.serverFriendlyDescription,
      identifier: this.adiIdentifier,
      localUUID: this.localUserUuid,
    };
  }

  static generateDefaults(): AnisetteDeviceConfig {
    return {
      serverFriendlyDescription: DEFAULT_CLIENT_INFO,
      uniqueDeviceId: randomUUID(),
      adiId: randomHex(8, false),
      localUserUuid: randomHex(32, true),
    };
  }
}
